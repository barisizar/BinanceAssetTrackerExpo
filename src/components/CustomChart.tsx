import React, {useState, useEffect, useRef, useMemo} from 'react';
import {
  View,
  ScrollView,
  Text,
  StyleSheet,
  Dimensions,
  PanResponder,
} from 'react-native';
import Svg, {Path, Circle, Text as SvgText, Line} from 'react-native-svg';
import { formatPriceInINR } from '../utils/formatPrice';

const {width: SCREEN_WIDTH} = Dimensions.get('window');
const CHART_HEIGHT = 250;
const CHART_WIDTH = SCREEN_WIDTH * 3;
const X_AXIS_HEIGHT = 30;
const MAX_X_LABELS = 12;
const Y_AXIS_PADDING = 0.1; // 10% padding

interface ChartData {
  labels: string[];
  datasets: {data: number[]}[];
}

interface CustomChartProps {
  data: ChartData;
  onSelectDataPoint: (label: string, value: number) => void;
}

const CustomChart: React.FC<CustomChartProps> = ({data, onSelectDataPoint}) => {
  const [selectedPoint, setSelectedPoint] = useState<{
    x: number;
    y: number;
    value: number;
    label: string;
  } | null>(null);
  const scrollViewRef = useRef<ScrollView>(null);

  const {minValue, maxValue, path, dataPoints} = useMemo(() => {
    if (!data.datasets[0].data.length) {
      return {minValue: 0, maxValue: 0, path: '', dataPoints: []};
    }

    const minValue = Math.min(...data.datasets[0].data);
    const maxValue = Math.max(...data.datasets[0].data);

    // add padding
    const range = maxValue - minValue;
    const paddedMin = minValue - range * Y_AXIS_PADDING;
    const paddedMax = maxValue + range * Y_AXIS_PADDING;

    const getY = (value: number) => {
      return (
        CHART_HEIGHT -
        ((value - paddedMin) / (paddedMax - paddedMin)) * CHART_HEIGHT
      );
    };

    const dataPoints = data.datasets[0].data.map((point, index) => ({
      x: (index / (data.datasets[0].data.length - 1)) * CHART_WIDTH,
      y: getY(point),
      value: point,
      label: data.labels[index],
    }));

    const path = dataPoints.reduce((acc, point, index) => {
      return `${acc} ${index === 0 ? 'M' : 'L'} ${point.x} ${point.y}`;
    }, '');

    return {minValue: paddedMin, maxValue: paddedMax, path, dataPoints};
  }, [data]);

  const visibleLabels = useMemo(() => {
    const totalPoints = dataPoints.length;
    let step;

    if (totalPoints <= MAX_X_LABELS) {
      step = 1;
    } else if (totalPoints <= MAX_X_LABELS * 2) {
      step = 2;
    } else {
      step = Math.ceil(totalPoints / MAX_X_LABELS);
    }

    return dataPoints.filter((_, index) => index % step === 0);
  }, [dataPoints]);

  const panResponder = PanResponder.create({
    onStartShouldSetPanResponder: () => true,
    onMoveShouldSetPanResponder: () => true,
    onPanResponderGrant: evt =>
      handleTouch(evt.nativeEvent.locationX, evt.nativeEvent.locationY),
    onPanResponderMove: evt =>
      handleTouch(evt.nativeEvent.locationX, evt.nativeEvent.locationY),
  });

  const handleTouch = (touchX: number, touchY: number) => {
    const scrollX = scrollViewRef.current?.contentOffset?.x || 0;
    const adjustedX = touchX + scrollX;

    const closestPoint = dataPoints.reduce((closest, point) => {
      return Math.abs(point.x - adjustedX) < Math.abs(closest.x - adjustedX)
        ? point
        : closest;
    });

    setSelectedPoint(closestPoint);
    onSelectDataPoint(closestPoint.label, closestPoint.value);
  };

  useEffect(() => {
    setTimeout(() => {
      scrollViewRef.current?.scrollToEnd({animated: false});
    }, 100);
  }, [data.labels]);

  if (
    !data ||
    !data.datasets ||
    data.datasets.length === 0 ||
    !data.datasets[0].data ||
    data.datasets[0].data.length === 0
  ) {
    return (
      <View style={styles.errorContainer}>
        <Text>No data available</Text>
      </View>
    );
  }

  return (
    <View testID="custom-chart" style={styles.container} >
      <ScrollView
        ref={scrollViewRef}
        horizontal
        showsHorizontalScrollIndicator={false}
        contentContainerStyle={styles.scrollViewContent}
        {...panResponder.panHandlers}>
        <Svg width={CHART_WIDTH} height={CHART_HEIGHT + X_AXIS_HEIGHT}>
          <Path d={path} fill="none" stroke="#4A90E2" strokeWidth="2" />
          {selectedPoint && (
            <>
              <Circle
                cx={selectedPoint.x}
                cy={selectedPoint.y}
                r="5"
                fill="#4A90E2"
              />
              <SvgText
                x={selectedPoint.x}
                y={selectedPoint.y - 10}
                fill="black"
                fontSize="12"
                fontWeight="bold"
                textAnchor="middle">
                {formatPriceInINR(selectedPoint.value)}
              </SvgText>
            </>
          )}
          {/* X-axis line */}
          <Line
            x1="0"
            y1={CHART_HEIGHT}
            x2={CHART_WIDTH}
            y2={CHART_HEIGHT}
            stroke="#E0E0E0"
            strokeWidth="1"
          />
          {/* X-axis labels and ticks */}
          {visibleLabels.map((point, index) => (
            <React.Fragment key={index}>
              <Line
                x1={point.x}
                y1={CHART_HEIGHT}
                x2={point.x}
                y2={CHART_HEIGHT + 5}
                stroke="#E0E0E0"
                strokeWidth="1"
              />
              <SvgText
                x={point.x}
                y={CHART_HEIGHT + 20}
                fontSize="10"
                fill="#666"
                textAnchor="middle">
                {point.label}
              </SvgText>
            </React.Fragment>
          ))}
        </Svg>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    height: CHART_HEIGHT + X_AXIS_HEIGHT,
    backgroundColor: 'white',
  },
  scrollViewContent: {
    width: CHART_WIDTH,
    height: CHART_HEIGHT + X_AXIS_HEIGHT,
  },
  errorContainer: {
    height: CHART_HEIGHT + X_AXIS_HEIGHT,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'white',
  },
});

export default CustomChart;
