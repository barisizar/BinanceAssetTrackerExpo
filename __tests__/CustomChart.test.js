import React from 'react';
import renderer from 'react-test-renderer';
import CustomChart from '../src/components/CustomChart';

jest.mock('react-native', () => {
    const RN = jest.requireActual('react-native');
    RN.Dimensions = {
        get: jest.fn().mockReturnValue({ width: 400, height: 800 }),
    };
    RN.PanResponder = {
        create: jest.fn(() => ({
            panHandlers: {},
        })),
    };
    return RN;
});

describe('CustomChart', () => {
    const mockData = {
        labels: ['Jan', 'Feb', 'Mar', 'Apr', 'May'],
        datasets: [{ data: [10, 20, 15, 25, 30] }],
    };

    it('renders correctly', () => {
        const tree = renderer.create(
            <CustomChart data={mockData} onSelectDataPoint={jest.fn()} />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
