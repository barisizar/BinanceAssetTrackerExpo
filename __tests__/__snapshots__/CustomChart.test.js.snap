// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomChart renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "height": 280,
    }
  }
  testID="custom-chart"
>
  <RCTScrollView
    contentContainerStyle={
      {
        "height": 280,
        "width": 2250,
      }
    }
    horizontal={true}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <RNSVGSvgView
        bbHeight={280}
        bbWidth={2250}
        focusable={false}
        height={280}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "flex": 0,
              "height": 280,
              "width": 2250,
            },
          ]
        }
        width={2250}
      >
        <RNSVGGroup
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
        >
          <RNSVGPath
            d=" M 0 229.16666666666666 L 562.5 125 L 1125 177.08333333333331 L 1687.5 72.91666666666666 L 2250 20.833333333333343"
            fill={null}
            propList={
              [
                "fill",
                "stroke",
                "strokeWidth",
              ]
            }
            stroke={
              {
                "payload": 4283076834,
                "type": 0,
              }
            }
            strokeWidth="2"
          />
          <RNSVGLine
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
              ]
            }
            stroke={
              {
                "payload": 4292927712,
                "type": 0,
              }
            }
            strokeWidth="1"
            x1="0"
            x2={2250}
            y1={250}
            y2={250}
          />
          <RNSVGLine
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
              ]
            }
            stroke={
              {
                "payload": 4292927712,
                "type": 0,
              }
            }
            strokeWidth="1"
            x1={0}
            x2={0}
            y1={250}
            y2={255}
          />
          <RNSVGText
            content={null}
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4284900966,
                "type": 0,
              }
            }
            font={
              {
                "fontSize": "10",
                "textAnchor": "middle",
              }
            }
            propList={
              [
                "fill",
              ]
            }
            rotate={[]}
            x={
              [
                0,
              ]
            }
            y={
              [
                270,
              ]
            }
          >
            <RNSVGTSpan
              content="Jan"
              dx={[]}
              dy={[]}
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              font={{}}
              rotate={[]}
              x={[]}
              y={[]}
            />
          </RNSVGText>
          <RNSVGLine
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
              ]
            }
            stroke={
              {
                "payload": 4292927712,
                "type": 0,
              }
            }
            strokeWidth="1"
            x1={562.5}
            x2={562.5}
            y1={250}
            y2={255}
          />
          <RNSVGText
            content={null}
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4284900966,
                "type": 0,
              }
            }
            font={
              {
                "fontSize": "10",
                "textAnchor": "middle",
              }
            }
            propList={
              [
                "fill",
              ]
            }
            rotate={[]}
            x={
              [
                562.5,
              ]
            }
            y={
              [
                270,
              ]
            }
          >
            <RNSVGTSpan
              content="Feb"
              dx={[]}
              dy={[]}
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              font={{}}
              rotate={[]}
              x={[]}
              y={[]}
            />
          </RNSVGText>
          <RNSVGLine
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
              ]
            }
            stroke={
              {
                "payload": 4292927712,
                "type": 0,
              }
            }
            strokeWidth="1"
            x1={1125}
            x2={1125}
            y1={250}
            y2={255}
          />
          <RNSVGText
            content={null}
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4284900966,
                "type": 0,
              }
            }
            font={
              {
                "fontSize": "10",
                "textAnchor": "middle",
              }
            }
            propList={
              [
                "fill",
              ]
            }
            rotate={[]}
            x={
              [
                1125,
              ]
            }
            y={
              [
                270,
              ]
            }
          >
            <RNSVGTSpan
              content="Mar"
              dx={[]}
              dy={[]}
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              font={{}}
              rotate={[]}
              x={[]}
              y={[]}
            />
          </RNSVGText>
          <RNSVGLine
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
              ]
            }
            stroke={
              {
                "payload": 4292927712,
                "type": 0,
              }
            }
            strokeWidth="1"
            x1={1687.5}
            x2={1687.5}
            y1={250}
            y2={255}
          />
          <RNSVGText
            content={null}
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4284900966,
                "type": 0,
              }
            }
            font={
              {
                "fontSize": "10",
                "textAnchor": "middle",
              }
            }
            propList={
              [
                "fill",
              ]
            }
            rotate={[]}
            x={
              [
                1687.5,
              ]
            }
            y={
              [
                270,
              ]
            }
          >
            <RNSVGTSpan
              content="Apr"
              dx={[]}
              dy={[]}
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              font={{}}
              rotate={[]}
              x={[]}
              y={[]}
            />
          </RNSVGText>
          <RNSVGLine
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
              ]
            }
            stroke={
              {
                "payload": 4292927712,
                "type": 0,
              }
            }
            strokeWidth="1"
            x1={2250}
            x2={2250}
            y1={250}
            y2={255}
          />
          <RNSVGText
            content={null}
            dx={[]}
            dy={[]}
            fill={
              {
                "payload": 4284900966,
                "type": 0,
              }
            }
            font={
              {
                "fontSize": "10",
                "textAnchor": "middle",
              }
            }
            propList={
              [
                "fill",
              ]
            }
            rotate={[]}
            x={
              [
                2250,
              ]
            }
            y={
              [
                270,
              ]
            }
          >
            <RNSVGTSpan
              content="May"
              dx={[]}
              dy={[]}
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              font={{}}
              rotate={[]}
              x={[]}
              y={[]}
            />
          </RNSVGText>
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </RCTScrollView>
</View>
`;
